File: Kakao/인턴_2019_크레인 인형뽑기 게임.cpp
// 출제의도: "구현, 스택", Time Complexity: O(m*b), Space Complexity: O(m). (m: moves.size, b: board.size)

#include <string>
#include <vector>
#include <stack>
#include <iostream>

using namespace std;

int solution(vector<vector<int>> board, vector<int> moves) {
    int answer = 0;
    stack<int> st;
    
    for(int i=0;i<moves.size();i++) {
        int select = moves[i]-1;
        for(int j=0;j<board.size();j++) {
            if(board[j][select]) {
                if(st.empty())  st.push(board[j][select]);
                else if(!st.empty() && st.top()==board[j][select]) {
                    st.pop();
                    answer += 2;
                }
                else {
                    st.push(board[j][select]);
                }
                
                board[j][select] = 0;
                break;
            }
        }
    }
    return answer;
}

File: Kakao/인턴_2019_크레인 인형뽑기 게임.swift
// "Stack", Time Complexity: O(mn), Space Complexity: O(n) (m: board.size, n: moves.size)
// reduce를 이용하여 간단하게 풀이.

import Foundation

var board = [[Int]]()
var moves = [Int]()

func pick(_ pos: Int) -> Int? {
    for i in 0..<board.count {
        if board[i][pos-1] > 0 {
            let doll = board[i][pos-1]
            board[i][pos-1] = 0
            return  doll
        }
    }
    return nil
}

func solution(_ param0:[[Int]], _ param1:[Int]) -> Int {
    var ans: Int = 0
    board = param0
    moves = param1
    
    moves.reduce(into: [Int]()) { stack, pos in
        guard let doll = pick(pos) else { return }
        if stack.isEmpty || stack.last != doll {
            stack.append(doll)
            return
        }
        stack.removeLast()
        ans += 2
    }
    return ans
}
